Avoid usage of __builtin_popcountll when -simdutf is unset

wz1000 opened the pull request:
Also avoid linking against gcc/gcc_s on all platforms.

This works around https://gitlab.haskell.org/ghc/ghc/-/issues/21787
and https://gitlab.haskell.org/ghc/ghc/-/issues/19900 which cause
problems when GHC's RTS linker tries to load `text`, which occurs if
you use a statically linked GHC to compile a file with a TH splice that
depends on `text`.

Since we don't require SSE4.2 to build `text -simdutf`, this shouldn't
be much of a pessimisation.

Fixes https://github.com/haskell/text/issues/450

This is meant to be a temporary workaround for https://gitlab.haskell.org/ghc/ghc/-/issues/21787 while we work on a robust method for properly exposing all GCC symbols from the RTS linker.

However, older versions of GHC (particularly the 9.0 series and earlier) which won't be patched still need a workaround so that `text-2.0` is usable under all configurations.

It is also problematic to include `extra-libraries: gcc_s` when compiling with clang.

Bodigrim commented:
Thanks @wz1000. First of all, before we discuss the chosen approach, I'd like to see a reproducible evidence of the issue as an additional CI job. This would help us to validate the solution and prevent future regressions.

wz1000 commented:
@Bodigrim I've modified the `-simdutf` CI job to test on windows (which is the only GHC distribution we ship with a statically linked GHC). I also modified it to test GHC-9.2.3 instead of `latest` because currently `latest` resolves to `9.2.2`, and chocolatey includes a workaround on that version which also masks the text issue.

See https://github.com/haskell/text/runs/7279337599 (which is a run from #454) for an example of the failure without this patch.

wz1000 commented:
I've improved the implementation to only use 2 bit shifts and a multiplication to compute the popcount and added a CI job that tests it on alpine.

wz1000 commented:
Unfortunately the static alpine GHC binaries are not usable because of https://gitlab.haskell.org/ghc/ghc/-/issues/21844

I'm stumped by this, not sure how to add a non windows CI job that demonstrates the problem.

wz1000 commented:
I guess I can try to run it in an alpine container and it should work. I'll try this tomorrow.

Bodigrim commented:
> I'm stumped by this, not sure how to add a non windows CI job that demonstrates the problem.

As I suggested in https://github.com/haskell/text/issues/450#issuecomment-1180871134, try to reproduce `bytestring` CI with respect to running Windows job with a clean PATH: https://github.com/haskell/bytestring/blob/22b36125ac52605e807b7b96ef31e8f087248f17/.github/workflows/ci.yml#L93-L99
If this proves that `extra-libraries: gcc_s` is not a good solution, we are all set, no need for Alpine.

wz1000 commented:
I've fixed the alpine job and also attempted to add a windows job to perform the check you added in bytestring, but that job seems to fail in a different way to the intended failure mode. See https://github.com/haskell/text/runs/7337239414?check_suite_focus=true

Bodigrim commented:
The simplest setup I can come up is https://github.com/Bodigrim/text/tree/purge-path. After the first commit the build succeeds, but as soon as we purge PATH, if fails (the error message is just error code 1, that's fine). If this setup is enough, I'd rather avoid Alpine job with unreleased GHC version. 

Now I don't quite understand what does it have to do with `simdutf` flag. The code is used unconditionally for any `length` / `drop` / `take`, so its performance is absolutely crucial. Please acompany your patch with benchmark results (`cabal bench --benchmark-options='-p length'`).

bgamari commented:
The failure in #454 is due not to `text` but rather `bytestring`, which `text` links against. See https://github.com/haskell/bytestring/issues/497.

This should be fixed in the `bytestring` shipped with 9.2.3.

bgamari commented:
#454 also neglects to disable the `simdutf8` flag, which introduces a dependency on `libstdc++`, which will of course fail for the same reason described in the last paragraph of [GHC #20878](https://gitlab.haskell.org/ghc/ghc/-/issues/20878).

Bodigrim commented:
@wz1000 could you please check performance impact of your change?

wz1000 commented:
Here are the results of running `cabal bench --benchmark-options='-p length'` with GHC 9.2.3

Before:

```
All
  Pure
    tiny
      length
        cons
          Text:     OK (0.22s)
            22.4 ns ± 1.7 ns
          LazyText: OK (0.82s)
            23.6 ns ± 1.4 ns
        decode
          Text:     OK (0.20s)
            41.2 ns ± 3.6 ns
          LazyText: OK (0.29s)
            124  ns ± 6.2 ns
        drop
          Text:     OK (0.17s)
            33.6 ns ± 3.1 ns
          LazyText: OK (0.18s)
            35.7 ns ± 3.3 ns
        filter
          Text:     OK (0.19s)
            18.5 ns ± 1.7 ns
          LazyText: OK (0.21s)
            40.3 ns ± 3.9 ns
        filter.filter
          Text:     OK (0.23s)
            23.4 ns ± 1.9 ns
          LazyText: OK (0.20s)
            40.3 ns ± 3.0 ns
        init
          Text:     OK (0.31s)
            30.3 ns ± 1.6 ns
          LazyText: OK (0.26s)
            26.6 ns ± 1.8 ns
        intercalate
          Text:     OK (0.29s)
            30.5 ns ± 2.2 ns
          LazyText: OK (0.15s)
            31.5 ns ± 2.8 ns
        intersperse
          Text:     OK (0.27s)
            29.2 ns ± 1.8 ns
          LazyText: OK (0.27s)
            27.9 ns ± 1.7 ns
        map
          Text:     OK (0.88s)
            24.4 ns ± 1.2 ns
          LazyText: OK (0.25s)
            25.2 ns ± 2.3 ns
        map.map
          Text:     OK (0.45s)
            24.1 ns ± 2.0 ns
          LazyText: OK (0.46s)
            24.9 ns ± 700 ps
        replicate char
          Text:     OK (0.40s)
            21.2 ns ± 1.0 ns
          LazyText: OK (0.18s)
            16.8 ns ± 1.4 ns
        replicate string
          Text:     OK (0.25s)
            25.3 ns ± 1.7 ns
          LazyText: OK (0.21s)
            21.3 ns ± 1.8 ns
        take
          Text:     OK (0.45s)
            24.4 ns ± 1.6 ns
          LazyText: OK (0.16s)
            29.1 ns ± 2.7 ns
        tail
          Text:     OK (0.24s)
            23.8 ns ± 1.9 ns
          LazyText: OK (0.25s)
            24.7 ns ± 1.8 ns
        toLower
          Text:     OK (0.27s)
            115  ns ± 8.4 ns
          LazyText: OK (0.24s)
            196  ns ±  13 ns
        toUpper
          Text:     OK (0.14s)
            118  ns ±  12 ns
          LazyText: OK (0.15s)
            228  ns ±  22 ns
        words
          Text:     OK (0.22s)
            21.5 ns ± 1.9 ns
          LazyText: OK (1.29s)
            36.4 ns ± 1.8 ns
        zipWith
          Text:     OK (0.25s)
            25.8 ns ± 2.2 ns
          LazyText: OK (0.25s)
            26.6 ns ± 1.3 ns
    ascii-small
      length
        cons
          Text:     OK (0.78s)
            11.7 μs ± 187 ns
          LazyText: OK (0.22s)
            11.8 μs ± 928 ns
        decode
          Text:     OK (0.67s)
            17.6 μs ± 1.4 μs
          LazyText: OK (0.72s)
            18.7 μs ± 793 ns
        drop
          Text:     OK (0.22s)
            11.9 μs ± 1.2 μs
          LazyText: OK (0.42s)
            11.9 μs ± 502 ns
        filter
          Text:     OK (0.30s)
            139  μs ± 8.3 μs
          LazyText: OK (0.17s)
            145  μs ±  11 μs
        filter.filter
          Text:     OK (0.13s)
            137  μs ±  11 μs
          LazyText: OK (0.17s)
            144  μs ±  11 μs
        init
          Text:     OK (0.20s)
            11.4 μs ± 670 ns
          LazyText: OK (0.22s)
            11.6 μs ± 695 ns
        intercalate
          Text:     OK (0.41s)
            23.2 μs ± 1.5 μs
          LazyText: OK (0.23s)
            25.4 μs ± 2.0 μs
        intersperse
          Text:     OK (0.41s)
            11.7 μs ± 432 ns
          LazyText: OK (0.22s)
            11.6 μs ± 797 ns
        map
          Text:     OK (0.22s)
            11.7 μs ± 698 ns
          LazyText: OK (0.42s)
            11.9 μs ± 363 ns
        map.map
          Text:     OK (0.22s)
            11.8 μs ± 718 ns
          LazyText: OK (0.22s)
            11.9 μs ± 898 ns
        replicate char
          Text:     OK (0.25s)
            23.8 ns ± 1.8 ns
          LazyText: OK (0.21s)
            19.2 ns ± 1.6 ns
        replicate string
          Text:     OK (0.29s)
            28.5 ns ± 1.9 ns
          LazyText: OK (0.25s)
            23.9 ns ± 1.9 ns
        take
          Text:     OK (0.16s)
            7.98 μs ± 669 ns
          LazyText: OK (0.15s)
            7.89 μs ± 699 ns
        tail
          Text:     OK (0.22s)
            11.8 μs ± 947 ns
          LazyText: OK (0.23s)
            11.8 μs ± 735 ns
        toLower
          Text:     OK (0.35s)
            1.30 ms ± 102 μs
          LazyText: OK (0.24s)
            1.74 ms ± 128 μs
        toUpper
          Text:     OK (0.24s)
            1.76 ms ±  91 μs
          LazyText: OK (0.16s)
            2.26 ms ± 203 μs
        words
          Text:     OK (0.18s)
            311  μs ±  27 μs
          LazyText: OK (0.34s)
            622  μs ±  26 μs
        zipWith
          Text:     OK (0.40s)
            11.5 μs ± 797 ns
          LazyText: OK (0.22s)
            11.7 μs ± 918 ns
    ascii
      length
        cons
          Text:     OK (0.56s)
            10.2 ms ± 911 μs
          LazyText: OK (3.44s)
            10.5 ms ± 857 μs
        decode
          Text:     OK (1.80s)
            18.1 ms ± 843 μs
          LazyText: OK (1.13s)
            18.9 ms ± 564 μs
        drop
          Text:     OK (0.95s)
            10.4 ms ± 480 μs
          LazyText: OK (0.46s)
            10.3 ms ± 934 μs
        filter
          Text:     OK (0.57s)
            122  ms ± 4.4 ms
          LazyText: OK (0.49s)
            128  ms ± 5.0 ms
        filter.filter
          Text:     OK (0.57s)
            121  ms ± 3.2 ms
          LazyText: OK (0.59s)
            129  ms ± 4.7 ms
        init
          Text:     OK (0.72s)
            13.6 ms ± 742 μs
          LazyText: OK (0.55s)
            10.2 ms ± 734 μs
        intercalate
          Text:     OK (0.64s)
            21.0 ms ± 1.2 ms
          LazyText: OK (0.36s)
            21.9 ms ± 1.6 ms
        intersperse
          Text:     OK (0.86s)
            11.2 ms ± 1.0 ms
          LazyText: OK (0.74s)
            10.8 ms ± 337 μs
        map
          Text:     OK (0.85s)
            10.5 ms ± 347 μs
          LazyText: OK (0.46s)
            10.3 ms ± 813 μs
        map.map
          Text:     OK (0.59s)
            10.0 ms ± 425 μs
          LazyText: OK (0.71s)
            10.4 ms ± 786 μs
        replicate char
          Text:     OK (2.47s)
            22.8 ns ± 1.8 ns
          LazyText: OK (2.41s)
            18.7 ns ± 1.8 ns
        replicate string
          Text:     OK (2.48s)
            27.3 ns ± 1.9 ns
          LazyText: OK (2.79s)
            23.2 ns ± 1.4 ns
        take
          Text:     OK (0.71s)
            6.69 ms ± 352 μs
          LazyText: OK (1.46s)
            6.93 ms ± 477 μs
        tail
          Text:     OK (0.56s)
            10.2 ms ± 814 μs
          LazyText: OK (0.72s)
            9.94 ms ± 656 μs
        toLower
          Text:     OK (8.66s)
            1.206 s ±  15 ms
          LazyText: OK (4.66s)
            1.485 s ±  25 ms
        toUpper
          Text:     OK (4.55s)
            1.479 s ±  30 ms
          LazyText: OK (5.70s)
            1.832 s ± 5.5 ms
        words
          Text:     OK (0.95s)
            252  ms ± 4.5 ms
          LazyText: OK (1.74s)
            515  ms ± 3.5 ms
        zipWith
          Text:     OK (0.44s)
            9.80 ms ± 773 μs
          LazyText: OK (0.45s)
            9.80 ms ± 907 μs
    english
      length
        cons
          Text:     OK (0.26s)
            643  μs ±  44 μs
          LazyText: OK (0.24s)
            650  μs ±  43 μs
        decode
          Text:     OK (1.12s)
            942  μs ±  23 μs
          LazyText: OK (0.69s)
            1.12 ms ±  66 μs
        drop
          Text:     OK (0.24s)
            655  μs ±  44 μs
          LazyText: OK (0.23s)
            654  μs ±  55 μs
        filter
          Text:     OK (0.28s)
            7.71 ms ± 386 μs
          LazyText: OK (0.15s)
            8.24 ms ± 822 μs
        filter.filter
          Text:     OK (0.15s)
            7.79 ms ± 706 μs
          LazyText: OK (0.30s)
            8.16 ms ± 388 μs
        init
          Text:     OK (0.23s)
            646  μs ±  48 μs
          LazyText: OK (0.24s)
            651  μs ±  46 μs
        intercalate
          Text:     OK (0.22s)
            1.32 ms ± 100 μs
          LazyText: OK (0.24s)
            1.39 ms ±  91 μs
        intersperse
          Text:     OK (0.14s)
            650  μs ±  59 μs
          LazyText: OK (0.24s)
            653  μs ±  59 μs
        map
          Text:     OK (0.24s)
            653  μs ±  57 μs
          LazyText: OK (0.41s)
            653  μs ±  36 μs
        map.map
          Text:     OK (0.41s)
            649  μs ±  30 μs
          LazyText: OK (0.42s)
            656  μs ±  63 μs
        replicate char
          Text:     OK (0.39s)
            22.9 ns ± 2.2 ns
          LazyText: OK (0.35s)
            18.8 ns ± 1.7 ns
        replicate string
          Text:     OK (0.42s)
            27.1 ns ± 2.1 ns
          LazyText: OK (12.71s)
            22.7 ns ± 212 ps
        take
          Text:     OK (0.29s)
            421  μs ±  27 μs
          LazyText: OK (0.48s)
            376  μs ±  17 μs
        tail
          Text:     OK (0.21s)
            572  μs ±  43 μs
          LazyText: OK (0.21s)
            574  μs ±  44 μs
        toLower
          Text:     OK (2.76s)
            90.6 ms ± 7.8 ms
          LazyText: OK (0.76s)
            102  ms ± 4.8 ms
        toUpper
          Text:     OK (0.36s)
            114  ms ± 3.4 ms
          LazyText: OK (0.44s)
            141  ms ± 9.3 ms
        words
          Text:     OK (0.32s)
            20.0 ms ± 776 μs
          LazyText: OK (2.54s)
            38.1 ms ± 903 μs
        zipWith
          Text:     OK (0.25s)
            688  μs ±  60 μs
          LazyText: OK (0.39s)
            589  μs ±  26 μs
    russian
      length
        cons
          Text:     OK (0.16s)
            1.91 μs ± 175 ns
          LazyText: OK (0.17s)
            2.00 μs ± 176 ns
        decode
          Text:     OK (1.17s)
            4.02 μs ± 240 ns
          LazyText: OK (1.11s)
            3.77 μs ±  54 ns
        drop
          Text:     OK (0.29s)
            1.93 μs ±  83 ns
          LazyText: OK (0.15s)
            1.90 μs ± 175 ns
        filter
          Text:     OK (0.20s)
            22.4 μs ± 1.8 μs
          LazyText: OK (0.25s)
            27.5 μs ± 1.9 μs
        filter.filter
          Text:     OK (0.38s)
            22.0 μs ± 1.3 μs
          LazyText: OK (0.13s)
            27.5 μs ± 2.7 μs
        init
          Text:     OK (0.14s)
            1.87 μs ± 172 ns
          LazyText: OK (0.51s)
            1.77 μs ± 118 ns
        intercalate
          Text:     OK (0.22s)
            2.84 μs ± 177 ns
          LazyText: OK (0.22s)
            2.98 μs ± 234 ns
        intersperse
          Text:     OK (0.26s)
            1.81 μs ±  98 ns
          LazyText: OK (0.27s)
            1.81 μs ± 104 ns
        map
          Text:     OK (0.27s)
            1.80 μs ± 105 ns
          LazyText: OK (0.27s)
            1.81 μs ± 111 ns
        map.map
          Text:     OK (0.27s)
            1.81 μs ± 132 ns
          LazyText: OK (0.16s)
            1.88 μs ± 174 ns
        replicate char
          Text:     OK (0.25s)
            23.0 ns ± 1.9 ns
          LazyText: OK (0.20s)
            18.4 ns ± 1.5 ns
        replicate string
          Text:     OK (0.50s)
            26.4 ns ± 1.1 ns
          LazyText: OK (3.41s)
            24.9 ns ± 2.2 ns
        take
          Text:     OK (0.22s)
            1.40 μs ± 105 ns
          LazyText: OK (0.23s)
            1.43 μs ±  88 ns
        tail
          Text:     OK (0.30s)
            2.05 μs ± 107 ns
          LazyText: OK (0.17s)
            2.02 μs ± 178 ns
        toLower
          Text:     OK (0.16s)
            129  μs ±  12 μs
          LazyText: OK (0.76s)
            180  μs ± 7.0 μs
        toUpper
          Text:     OK (0.20s)
            177  μs ±  14 μs
          LazyText: OK (0.51s)
            229  μs ±  12 μs
        words
          Text:     OK (0.73s)
            41.6 μs ± 1.3 μs
          LazyText: OK (0.20s)
            82.8 μs ± 8.1 μs
        zipWith
          Text:     OK (4.05s)
            1.89 μs ±  53 ns
          LazyText: OK (0.14s)
            1.85 μs ± 171 ns
    japanese
      length
        cons
          Text:     OK (0.16s)
            1.98 μs ± 181 ns
          LazyText: OK (0.16s)
            1.97 μs ± 165 ns
        decode
          Text:     OK (1.17s)
            4.08 μs ±  70 ns
          LazyText: OK (1.19s)
            4.11 μs ±  73 ns
        drop
          Text:     OK (1.06s)
            1.90 μs ± 129 ns
          LazyText: OK (0.26s)
            1.87 μs ± 122 ns
        filter
          Text:     OK (0.21s)
            11.8 μs ± 1.1 μs
          LazyText: OK (0.27s)
            15.6 μs ± 892 ns
        filter.filter
          Text:     OK (0.83s)
            12.2 μs ± 554 ns
          LazyText: OK (0.27s)
            15.9 μs ± 919 ns
        init
          Text:     OK (1.14s)
            2.19 μs ± 183 ns
          LazyText: OK (1.13s)
            2.08 μs ±  49 ns
        intercalate
          Text:     OK (0.33s)
            4.46 μs ± 199 ns
          LazyText: OK (0.20s)
            4.98 μs ± 377 ns
        intersperse
          Text:     OK (0.30s)
            2.01 μs ± 107 ns
          LazyText: OK (0.62s)
            2.29 μs ± 180 ns
        map
          Text:     OK (0.61s)
            2.12 μs ± 107 ns
          LazyText: OK (0.64s)
            2.28 μs ± 110 ns
        map.map
          Text:     OK (0.18s)
            2.08 μs ± 182 ns
          LazyText: OK (0.31s)
            2.08 μs ± 198 ns
        replicate char
          Text:     OK (0.46s)
            24.4 ns ± 890 ps
          LazyText: OK (0.72s)
            19.7 ns ± 1.0 ns
        replicate string
          Text:     OK (0.29s)
            28.2 ns ± 1.4 ns
          LazyText: OK (0.25s)
            23.0 ns ± 2.1 ns
        take
          Text:     OK (0.23s)
            1.44 μs ±  85 ns
          LazyText: OK (0.21s)
            1.44 μs ± 139 ns
        tail
          Text:     OK (0.30s)
            2.11 μs ±  90 ns
          LazyText: OK (0.31s)
            2.08 μs ± 182 ns
        toLower
          Text:     OK (0.37s)
            83.3 μs ± 4.4 μs
          LazyText: OK (0.27s)
            116  μs ± 6.1 μs
        toUpper
          Text:     OK (0.70s)
            82.3 μs ± 5.6 μs
          LazyText: OK (0.46s)
            110  μs ± 4.6 μs
        words
          Text:     OK (0.15s)
            59.1 μs ± 5.3 μs
          LazyText: OK (0.20s)
            42.5 μs ± 4.2 μs
        zipWith
          Text:     OK (0.29s)
            1.98 μs ± 110 ns
          LazyText: OK (0.57s)
            2.09 μs ± 204 ns

All 216 tests passed (149.07s)
Benchmark text-benchmarks: FINISH
```

After:

```
Running 1 benchmarks...
Benchmark text-benchmarks: RUNNING...
All
  Pure
    tiny
      length
        cons
          Text:     OK (0.23s)
            23.3 ns ± 1.4 ns
          LazyText: OK (0.23s)
            24.6 ns ± 1.6 ns
        decode
          Text:     OK (0.20s)
            42.0 ns ± 3.4 ns
          LazyText: OK (0.28s)
            126  ns ± 8.0 ns
        drop
          Text:     OK (0.15s)
            32.1 ns ± 2.6 ns
          LazyText: OK (0.17s)
            32.8 ns ± 2.6 ns
        filter
          Text:     OK (0.31s)
            16.2 ns ± 956 ps
          LazyText: OK (0.27s)
            27.9 ns ± 1.4 ns
        filter.filter
          Text:     OK (0.16s)
            16.3 ns ± 1.5 ns
          LazyText: OK (0.14s)
            28.4 ns ± 2.6 ns
        init
          Text:     OK (0.19s)
            20.2 ns ± 1.7 ns
          LazyText: OK (0.24s)
            24.6 ns ± 1.3 ns
        intercalate
          Text:     OK (0.14s)
            26.3 ns ± 2.6 ns
          LazyText: OK (0.14s)
            28.2 ns ± 2.6 ns
        intersperse
          Text:     OK (0.13s)
            26.9 ns ± 2.6 ns
          LazyText: OK (0.26s)
            26.7 ns ± 1.8 ns
        map
          Text:     OK (0.45s)
            24.6 ns ± 1.9 ns
          LazyText: OK (0.26s)
            27.3 ns ± 1.9 ns
        map.map
          Text:     OK (0.25s)
            24.8 ns ± 1.5 ns
          LazyText: OK (0.26s)
            26.2 ns ± 1.3 ns
        replicate char
          Text:     OK (0.44s)
            24.1 ns ± 1.3 ns
          LazyText: OK (0.66s)
            18.4 ns ± 650 ps
        replicate string
          Text:     OK (0.25s)
            24.7 ns ± 1.4 ns
          LazyText: OK (0.22s)
            20.7 ns ± 1.9 ns
        take
          Text:     OK (0.24s)
            23.1 ns ± 1.6 ns
          LazyText: OK (0.28s)
            28.0 ns ± 1.6 ns
        tail
          Text:     OK (0.25s)
            25.0 ns ± 2.2 ns
          LazyText: OK (0.26s)
            25.9 ns ± 1.5 ns
        toLower
          Text:     OK (0.29s)
            117  ns ± 8.2 ns
          LazyText: OK (0.26s)
            208  ns ±  19 ns
        toUpper
          Text:     OK (0.17s)
            121  ns ±  11 ns
          LazyText: OK (0.29s)
            237  ns ±  17 ns
        words
          Text:     OK (0.21s)
            21.6 ns ± 1.4 ns
          LazyText: OK (0.35s)
            36.3 ns ± 2.1 ns
        zipWith
          Text:     OK (0.27s)
            26.7 ns ± 2.4 ns
          LazyText: OK (0.27s)
            28.5 ns ± 1.6 ns
    ascii-small
      length
        cons
          Text:     OK (0.21s)
            22.9 μs ± 2.2 μs
          LazyText: OK (0.20s)
            22.9 μs ± 1.5 μs
        decode
          Text:     OK (1.01s)
            28.5 μs ± 1.8 μs
          LazyText: OK (1.06s)
            29.2 μs ± 1.8 μs
        drop
          Text:     OK (0.22s)
            23.1 μs ± 1.5 μs
          LazyText: OK (0.20s)
            23.0 μs ± 1.4 μs
        filter
          Text:     OK (0.16s)
            136  μs ±  11 μs
          LazyText: OK (0.17s)
            144  μs ±  11 μs
        filter.filter
          Text:     OK (0.13s)
            137  μs ±  11 μs
          LazyText: OK (0.17s)
            144  μs ±  12 μs
        init
          Text:     OK (0.20s)
            22.7 μs ± 1.3 μs
          LazyText: OK (0.21s)
            23.0 μs ± 1.6 μs
        intercalate
          Text:     OK (0.16s)
            33.0 μs ± 2.8 μs
          LazyText: OK (0.17s)
            34.9 μs ± 2.7 μs
        intersperse
          Text:     OK (0.21s)
            22.7 μs ± 1.4 μs
          LazyText: OK (0.22s)
            23.1 μs ± 1.5 μs
        map
          Text:     OK (0.20s)
            21.5 μs ± 1.4 μs
          LazyText: OK (0.20s)
            21.6 μs ± 1.5 μs
        map.map
          Text:     OK (0.20s)
            21.3 μs ± 1.4 μs
          LazyText: OK (0.21s)
            21.8 μs ± 1.5 μs
        replicate char
          Text:     OK (0.23s)
            21.9 ns ± 1.3 ns
          LazyText: OK (0.19s)
            17.5 ns ± 1.4 ns
        replicate string
          Text:     OK (0.25s)
            24.2 ns ± 1.4 ns
          LazyText: OK (0.22s)
            20.2 ns ± 1.6 ns
        take
          Text:     OK (0.26s)
            14.1 μs ± 724 ns
          LazyText: OK (0.26s)
            14.2 μs ± 710 ns
        tail
          Text:     OK (0.21s)
            21.8 μs ± 1.5 μs
          LazyText: OK (0.21s)
            21.8 μs ± 1.5 μs
        toLower
          Text:     OK (0.17s)
            1.26 ms ±  85 μs
          LazyText: OK (0.22s)
            1.67 ms ±  91 μs
        toUpper
          Text:     OK (0.23s)
            1.66 ms ±  97 μs
          LazyText: OK (0.28s)
            2.10 ms ± 112 μs
        words
          Text:     OK (0.17s)
            297  μs ±  25 μs
          LazyText: OK (0.16s)
            622  μs ±  58 μs
        zipWith
          Text:     OK (0.20s)
            22.7 μs ± 1.5 μs
          LazyText: OK (0.20s)
            23.0 μs ± 1.6 μs
    ascii
      length
        cons
          Text:     OK (0.68s)
            18.5 ms ± 766 μs
          LazyText: OK (1.72s)
            19.0 ms ± 460 μs
        decode
          Text:     OK (1.37s)
            26.3 ms ± 1.7 ms
          LazyText: OK (4.31s)
            25.7 ms ± 568 μs
        drop
          Text:     OK (0.70s)
            19.3 ms ± 1.5 ms
          LazyText: OK (0.33s)
            18.9 ms ± 1.5 ms
        filter
          Text:     OK (0.55s)
            116  ms ± 8.6 ms
          LazyText: OK (4.29s)
            123  ms ± 3.3 ms
        filter.filter
          Text:     OK (0.33s)
            116  ms ± 5.0 ms
          LazyText: OK (0.56s)
            121  ms ± 3.6 ms
        init
          Text:     OK (0.43s)
            18.5 ms ± 1.8 ms
          LazyText: OK (0.43s)
            18.7 ms ± 1.6 ms
        intercalate
          Text:     OK (0.50s)
            27.9 ms ± 1.7 ms
          LazyText: OK (0.41s)
            28.4 ms ± 1.4 ms
        intersperse
          Text:     OK (0.74s)
            21.2 ms ± 1.7 ms
          LazyText: OK (1.91s)
            19.9 ms ± 207 μs
        map
          Text:     OK (0.70s)
            19.2 ms ± 1.9 ms
          LazyText: OK (1.06s)
            19.6 ms ± 486 μs
        map.map
          Text:     OK (0.45s)
            20.2 ms ± 1.8 ms
          LazyText: OK (0.47s)
            20.5 ms ± 1.5 ms
        replicate char
          Text:     OK (3.01s)
            28.8 ns ± 2.8 ns
          LazyText: OK (3.14s)
            19.4 ns ± 944 ps
        replicate string
          Text:     OK (3.69s)
            27.2 ns ± 2.4 ns
          LazyText: OK (2.67s)
            22.8 ns ± 2.3 ns
        take
          Text:     OK (0.64s)
            13.6 ms ± 723 μs
          LazyText: OK (0.87s)
            13.2 ms ± 1.1 ms
        tail
          Text:     OK (0.49s)
            22.0 ms ± 2.1 ms
          LazyText: OK (0.73s)
            21.1 ms ± 1.7 ms
        toLower
          Text:     OK (8.80s)
            1.201 s ±  41 ms
          LazyText: OK (4.86s)
            1.592 s ± 121 ms
        toUpper
          Text:     OK (4.66s)
            1.482 s ±  40 ms
          LazyText: OK (5.82s)
            1.867 s ±  29 ms
        words
          Text:     OK (0.98s)
            257  ms ± 3.7 ms
          LazyText: OK (1.80s)
            531  ms ±  17 ms
        zipWith
          Text:     OK (1.02s)
            19.2 ms ± 414 μs
          LazyText: OK (0.36s)
            19.4 ms ± 1.5 ms
    english
      length
        cons
          Text:     OK (0.80s)
            1.42 ms ±  33 μs
          LazyText: OK (0.80s)
            1.29 ms ±  78 μs
        decode
          Text:     OK (0.33s)
            1.93 ms ± 127 μs
          LazyText: OK (0.57s)
            1.83 ms ±  44 μs
        drop
          Text:     OK (0.23s)
            1.33 ms ± 106 μs
          LazyText: OK (0.22s)
            1.31 ms ± 119 μs
        filter
          Text:     OK (0.29s)
            7.81 ms ± 477 μs
          LazyText: OK (0.28s)
            7.98 ms ± 443 μs
        filter.filter
          Text:     OK (0.15s)
            7.59 ms ± 740 μs
          LazyText: OK (0.15s)
            7.98 ms ± 743 μs
        init
          Text:     OK (0.22s)
            1.25 ms ± 116 μs
          LazyText: OK (0.40s)
            1.31 ms ±  73 μs
        intercalate
          Text:     OK (0.28s)
            1.85 ms ± 126 μs
          LazyText: OK (0.30s)
            1.90 ms ± 110 μs
        intersperse
          Text:     OK (0.14s)
            1.27 ms ±  90 μs
          LazyText: OK (0.41s)
            1.32 ms ± 104 μs
        map
          Text:     OK (0.23s)
            1.30 ms ± 123 μs
          LazyText: OK (1.52s)
            1.36 ms ±  66 μs
        map.map
          Text:     OK (0.34s)
            2.00 ms ± 123 μs
          LazyText: OK (0.60s)
            1.91 ms ±  82 μs
        replicate char
          Text:     OK (1.03s)
            22.7 ns ± 592 ps
          LazyText: OK (0.36s)
            18.6 ns ± 1.6 ns
        replicate string
          Text:     OK (0.40s)
            24.0 ns ± 1.4 ns
          LazyText: OK (0.86s)
            19.4 ns ± 1.0 ns
        take
          Text:     OK (0.28s)
            828  μs ±  69 μs
          LazyText: OK (0.26s)
            847  μs ±  57 μs
        tail
          Text:     OK (0.23s)
            1.32 ms ± 121 μs
          LazyText: OK (0.22s)
            1.26 ms ± 121 μs
        toLower
          Text:     OK (0.58s)
            79.2 ms ± 5.0 ms
          LazyText: OK (0.75s)
            104  ms ± 4.7 ms
        toUpper
          Text:     OK (0.31s)
            98.1 ms ± 5.0 ms
          LazyText: OK (1.99s)
            130  ms ± 9.6 ms
        words
          Text:     OK (0.53s)
            16.1 ms ± 929 μs
          LazyText: OK (0.54s)
            34.8 ms ± 1.8 ms
        zipWith
          Text:     OK (0.39s)
            1.26 ms ±  62 μs
          LazyText: OK (0.42s)
            1.28 ms ±  83 μs
    russian
      length
        cons
          Text:     OK (0.14s)
            3.71 μs ± 349 ns
          LazyText: OK (0.27s)
            3.72 μs ± 205 ns
        decode
          Text:     OK (0.43s)
            5.81 μs ± 265 ns
          LazyText: OK (0.44s)
            5.62 μs ± 367 ns
        drop
          Text:     OK (0.28s)
            3.72 μs ± 242 ns
          LazyText: OK (1.16s)
            4.48 μs ± 306 ns
        filter
          Text:     OK (0.45s)
            25.8 μs ± 1.6 μs
          LazyText: OK (0.51s)
            29.0 μs ± 2.9 μs
        filter.filter
          Text:     OK (0.41s)
            23.9 μs ± 666 ns
          LazyText: OK (0.26s)
            29.2 μs ± 1.9 μs
        init
          Text:     OK (0.17s)
            4.03 μs ± 376 ns
          LazyText: OK (0.29s)
            3.89 μs ± 323 ns
        intercalate
          Text:     OK (0.18s)
            4.63 μs ± 370 ns
          LazyText: OK (0.17s)
            4.73 μs ± 398 ns
        intersperse
          Text:     OK (0.14s)
            3.80 μs ± 345 ns
          LazyText: OK (0.55s)
            3.97 μs ± 297 ns
        map
          Text:     OK (0.16s)
            4.01 μs ± 360 ns
          LazyText: OK (0.16s)
            3.98 μs ± 366 ns
        map.map
          Text:     OK (0.17s)
            4.03 μs ± 366 ns
          LazyText: OK (0.16s)
            3.98 μs ± 357 ns
        replicate char
          Text:     OK (1.67s)
            23.8 ns ± 502 ps
          LazyText: OK (0.37s)
            18.8 ns ± 1.0 ns
        replicate string
          Text:     OK (0.48s)
            25.1 ns ± 2.3 ns
          LazyText: OK (0.22s)
            20.8 ns ± 1.4 ns
        take
          Text:     OK (0.21s)
            2.69 μs ± 265 ns
          LazyText: OK (0.38s)
            2.69 μs ± 214 ns
        tail
          Text:     OK (0.29s)
            4.17 μs ± 346 ns
          LazyText: OK (0.17s)
            4.12 μs ± 357 ns
        toLower
          Text:     OK (1.07s)
            127  μs ±  10 μs
          LazyText: OK (0.20s)
            169  μs ±  16 μs
        toUpper
          Text:     OK (0.19s)
            167  μs ±  11 μs
          LazyText: OK (0.26s)
            237  μs ±  20 μs
        words
          Text:     OK (0.66s)
            36.6 μs ± 3.6 μs
          LazyText: OK (0.32s)
            70.9 μs ± 4.1 μs
        zipWith
          Text:     OK (0.28s)
            3.72 μs ± 206 ns
          LazyText: OK (0.27s)
            3.83 μs ± 170 ns
    japanese
      length
        cons
          Text:     OK (0.17s)
            4.09 μs ± 378 ns
          LazyText: OK (0.15s)
            4.07 μs ± 355 ns
        decode
          Text:     OK (0.89s)
            6.26 μs ± 150 ns
          LazyText: OK (0.91s)
            6.22 μs ± 520 ns
        drop
          Text:     OK (0.85s)
            6.02 μs ± 208 ns
          LazyText: OK (0.65s)
            4.21 μs ± 102 ns
        filter
          Text:     OK (0.24s)
            12.6 μs ± 675 ns
          LazyText: OK (0.15s)
            16.4 μs ± 1.5 μs
        filter.filter
          Text:     OK (0.24s)
            13.0 μs ± 662 ns
          LazyText: OK (0.15s)
            16.3 μs ± 1.5 μs
        init
          Text:     OK (0.17s)
            4.09 μs ± 396 ns
          LazyText: OK (0.16s)
            4.03 μs ± 332 ns
        intercalate
          Text:     OK (0.22s)
            6.07 μs ± 398 ns
          LazyText: OK (0.24s)
            6.46 μs ± 352 ns
        intersperse
          Text:     OK (0.16s)
            4.07 μs ± 403 ns
          LazyText: OK (0.17s)
            4.08 μs ± 332 ns
        map
          Text:     OK (0.17s)
            4.08 μs ± 339 ns
          LazyText: OK (0.17s)
            4.07 μs ± 377 ns
        map.map
          Text:     OK (0.17s)
            4.14 μs ± 337 ns
          LazyText: OK (0.17s)
            4.08 μs ± 336 ns
        replicate char
          Text:     OK (0.23s)
            21.9 ns ± 1.5 ns
          LazyText: OK (0.20s)
            17.8 ns ± 1.4 ns
        replicate string
          Text:     OK (0.26s)
            24.5 ns ± 1.4 ns
          LazyText: OK (0.22s)
            20.6 ns ± 1.6 ns
        take
          Text:     OK (0.22s)
            2.83 μs ± 239 ns
          LazyText: OK (0.21s)
            2.86 μs ± 213 ns
        tail
          Text:     OK (0.32s)
            4.32 μs ± 231 ns
          LazyText: OK (0.32s)
            4.37 μs ± 189 ns
        toLower
          Text:     OK (0.34s)
            76.7 μs ± 3.3 μs
          LazyText: OK (0.25s)
            109  μs ± 5.3 μs
        toUpper
          Text:     OK (0.33s)
            72.9 μs ± 4.0 μs
          LazyText: OK (0.24s)
            106  μs ± 6.6 μs
        words
          Text:     OK (0.23s)
            52.2 μs ± 4.3 μs
          LazyText: OK (0.25s)
            45.6 μs ± 2.7 μs
        zipWith
          Text:     OK (0.28s)
            3.71 μs ± 214 ns
          LazyText: OK (0.14s)
            3.79 μs ± 336 ns

All 216 tests passed (129.26s)
Benchmark text-benchmarks: FINISH
```

wz1000 commented:
Tried to improve the implementation of `popcount16`. Results at https://gist.github.com/wz1000/7604ab1663d8612bbd8485b13ae22f86#file-after-improved-csv.

```
All
  Pure
    tiny
      length
        cons
          Text:     OK (0.22s)
            21.4 ns ± 1.3 ns
          LazyText: OK (0.23s)
            23.5 ns ± 1.5 ns,  9% faster than baseline
        decode
          Text:     OK (0.21s)
            40.7 ns ± 3.2 ns
          LazyText: OK (0.29s)
            118  ns ± 5.4 ns, 10% faster than baseline
        drop
          Text:     OK (0.29s)
            29.6 ns ± 1.8 ns, 12% faster than baseline
          LazyText: OK (0.31s)
            31.5 ns ± 2.1 ns,  9% faster than baseline
        filter
          Text:     OK (0.17s)
            15.3 ns ± 1.4 ns,  8% faster than baseline
          LazyText: OK (0.27s)
            26.7 ns ± 1.7 ns,  6% faster than baseline
        filter.filter
          Text:     OK (0.29s)
            15.2 ns ± 692 ps,  8% faster than baseline
          LazyText: OK (0.50s)
            27.3 ns ± 2.2 ns
        init
          Text:     OK (0.36s)
            18.8 ns ± 1.2 ns, 11% faster than baseline
          LazyText: OK (0.25s)
            24.6 ns ± 2.0 ns,  8% faster than baseline
        intercalate
          Text:     OK (0.26s)
            25.2 ns ± 1.6 ns, 14% faster than baseline
          LazyText: OK (0.26s)
            26.0 ns ± 1.5 ns, 10% faster than baseline
        intersperse
          Text:     OK (0.26s)
            25.4 ns ± 2.0 ns, 10% faster than baseline
          LazyText: OK (0.27s)
            26.0 ns ± 2.3 ns, 15% faster than baseline
        map
          Text:     OK (0.23s)
            22.2 ns ± 2.2 ns, 13% faster than baseline
          LazyText: OK (0.25s)
            24.5 ns ± 2.1 ns, 18% faster than baseline
        map.map
          Text:     OK (0.43s)
            22.7 ns ± 1.0 ns, 11% faster than baseline
          LazyText: OK (0.25s)
            24.5 ns ± 2.2 ns, 10% faster than baseline
        replicate char
          Text:     OK (0.21s)
            19.7 ns ± 1.7 ns, 20% faster than baseline
          LazyText: OK (0.18s)
            15.7 ns ± 1.4 ns, 18% faster than baseline
        replicate string
          Text:     OK (0.24s)
            23.1 ns ± 1.7 ns, 20% faster than baseline
          LazyText: OK (0.20s)
            18.6 ns ± 1.3 ns, 23% faster than baseline
        take
          Text:     OK (0.23s)
            21.5 ns ± 2.1 ns, 20% faster than baseline
          LazyText: OK (0.26s)
            26.1 ns ± 1.8 ns, 20% faster than baseline
        tail
          Text:     OK (0.23s)
            21.9 ns ± 1.8 ns, 15% faster than baseline
          LazyText: OK (0.27s)
            26.3 ns ± 1.4 ns,  8% faster than baseline
        toLower
          Text:     OK (0.26s)
            103  ns ± 5.9 ns, 15% faster than baseline
          LazyText: OK (0.21s)
            183  ns ±  13 ns, 16% faster than baseline
        toUpper
          Text:     OK (0.99s)
            113  ns ± 3.0 ns, 14% faster than baseline
          LazyText: OK (0.27s)
            214  ns ±  11 ns, 12% faster than baseline
        words
          Text:     OK (0.40s)
            20.8 ns ± 1.0 ns, 11% faster than baseline
          LazyText: OK (0.18s)
            35.1 ns ± 3.0 ns, 12% faster than baseline
        zipWith
          Text:     OK (0.46s)
            25.4 ns ± 1.8 ns, 13% faster than baseline
          LazyText: OK (0.18s)
            32.2 ns ± 3.0 ns
    ascii-small
      length
        cons
          Text:     OK (0.32s)
            9.26 μs ± 716 ns, 23% faster than baseline
          LazyText: OK (0.18s)
            9.52 μs ± 920 ns, 22% faster than baseline
        decode
          Text:     OK (0.58s)
            15.0 μs ± 1.4 μs, 16% faster than baseline
          LazyText: OK (0.60s)
            15.5 μs ± 906 ns, 22% faster than baseline
        drop
          Text:     OK (0.33s)
            9.61 μs ± 772 ns, 13% faster than baseline
          LazyText: OK (0.19s)
            9.59 μs ± 668 ns, 19% faster than baseline
        filter
          Text:     OK (0.16s)
            133  μs ±  12 μs
          LazyText: OK (0.30s)
            136  μs ±  13 μs
        filter.filter
          Text:     OK (0.16s)
            131  μs ±  11 μs
          LazyText: OK (0.16s)
            137  μs ±  11 μs
        init
          Text:     OK (0.17s)
            9.22 μs ± 687 ns,  9% faster than baseline
          LazyText: OK (0.19s)
            9.56 μs ± 797 ns
        intercalate
          Text:     OK (0.19s)
            19.7 μs ± 1.4 μs, 10% faster than baseline
          LazyText: OK (0.40s)
            22.2 μs ± 1.0 μs, 13% faster than baseline
        intersperse
          Text:     OK (0.64s)
            9.15 μs ± 306 ns,  9% faster than baseline
          LazyText: OK (0.19s)
            9.41 μs ± 688 ns
        map
          Text:     OK (0.18s)
            9.16 μs ± 719 ns, 21% faster than baseline
          LazyText: OK (0.19s)
            9.37 μs ± 809 ns, 10% faster than baseline
        map.map
          Text:     OK (0.18s)
            9.09 μs ± 775 ns, 11% faster than baseline
          LazyText: OK (0.18s)
            8.96 μs ± 781 ns, 14% faster than baseline
        replicate char
          Text:     OK (0.22s)
            19.6 ns ± 1.9 ns, 10% faster than baseline
          LazyText: OK (0.32s)
            16.0 ns ± 862 ps, 19% faster than baseline
        replicate string
          Text:     OK (0.25s)
            23.6 ns ± 1.5 ns
          LazyText: OK (0.22s)
            19.3 ns ± 1.5 ns, 15% faster than baseline
        take
          Text:     OK (0.24s)
            6.17 μs ± 337 ns, 21% faster than baseline
          LazyText: OK (0.24s)
            6.23 μs ± 427 ns, 22% faster than baseline
        tail
          Text:     OK (0.18s)
            9.76 μs ± 844 ns, 17% faster than baseline
          LazyText: OK (0.19s)
            9.88 μs ± 814 ns, 13% faster than baseline
        toLower
          Text:     OK (0.32s)
            1.15 ms ±  82 μs, 14% faster than baseline
          LazyText: OK (0.22s)
            1.57 ms ±  97 μs, 11% faster than baseline
        toUpper
          Text:     OK (0.22s)
            1.63 ms ± 133 μs
          LazyText: OK (0.28s)
            2.04 ms ± 114 μs
        words
          Text:     OK (0.16s)
            271  μs ±  27 μs
          LazyText: OK (0.17s)
            573  μs ±  44 μs
        zipWith
          Text:     OK (0.64s)
            9.21 μs ± 374 ns, 16% faster than baseline
          LazyText: OK (0.17s)
            9.52 μs ± 697 ns, 16% faster than baseline
    ascii
      length
        cons
          Text:     OK (0.67s)
            7.83 ms ± 535 μs, 23% faster than baseline
          LazyText: OK (0.37s)
            8.05 ms ± 739 μs, 18% faster than baseline
        decode
          Text:     OK (1.54s)
            15.2 ms ± 1.4 ms, 12% faster than baseline
          LazyText: OK (1.49s)
            14.8 ms ± 562 μs, 20% faster than baseline
        drop
          Text:     OK (0.68s)
            8.08 ms ± 416 μs, 10% faster than baseline
          LazyText: OK (0.39s)
            8.21 ms ± 711 μs, 10% faster than baseline
        filter
          Text:     OK (0.99s)
            105  ms ± 2.5 ms,  6% faster than baseline
          LazyText: OK (0.49s)
            108  ms ± 5.7 ms, 18% faster than baseline
        filter.filter
          Text:     OK (0.29s)
            104  ms ± 3.3 ms,  9% faster than baseline
          LazyText: OK (0.51s)
            112  ms ± 7.1 ms
        init
          Text:     OK (0.28s)
            8.06 ms ± 718 μs, 22% faster than baseline
          LazyText: OK (0.39s)
            8.29 ms ± 692 μs, 21% faster than baseline
        intercalate
          Text:     OK (0.38s)
            16.9 ms ± 1.5 ms, 22% faster than baseline
          LazyText: OK (0.53s)
            17.6 ms ± 809 μs, 13% faster than baseline
        intersperse
          Text:     OK (0.47s)
            8.06 ms ± 756 μs, 19% faster than baseline
          LazyText: OK (0.72s)
            8.47 ms ± 363 μs, 19% faster than baseline
        map
          Text:     OK (0.48s)
            8.24 ms ± 769 μs, 15% faster than baseline
          LazyText: OK (0.40s)
            8.43 ms ± 704 μs, 12% faster than baseline
        map.map
          Text:     OK (0.66s)
            7.75 ms ± 493 μs, 15% faster than baseline
          LazyText: OK (0.92s)
            7.93 ms ± 394 μs, 20% faster than baseline
        replicate char
          Text:     OK (2.21s)
            20.0 ns ± 1.3 ns
          LazyText: OK (2.72s)
            16.1 ns ± 688 ps
        replicate string
          Text:     OK (2.56s)
            23.3 ns ± 1.0 ns
          LazyText: OK (3.45s)
            17.9 ns ± 212 ps, 28% faster than baseline
        take
          Text:     OK (0.58s)
            5.07 ms ± 396 μs, 26% faster than baseline
          LazyText: OK (0.50s)
            5.34 ms ± 440 μs, 28% faster than baseline
        tail
          Text:     OK (0.39s)
            8.46 ms ± 841 μs, 14% faster than baseline
          LazyText: OK (0.39s)
            8.27 ms ± 722 μs, 22% faster than baseline
        toLower
          Text:     OK (3.14s)
            1.016 s ± 3.2 ms, 22% faster than baseline
          LazyText: OK (4.11s)
            1.347 s ±  71 ms, 23% faster than baseline
        toUpper
          Text:     OK (4.17s)
            1.329 s ±  32 ms, 17% faster than baseline
          LazyText: OK (5.19s)
            1.700 s ±  42 ms, 14% faster than baseline
        words
          Text:     OK (0.84s)
            225  ms ± 8.9 ms, 11% faster than baseline
          LazyText: OK (1.64s)
            488  ms ±  14 ms, 18% faster than baseline
        zipWith
          Text:     OK (0.60s)
            7.91 ms ± 377 μs, 24% faster than baseline
          LazyText: OK (0.61s)
            8.37 ms ± 757 μs, 19% faster than baseline
    english
      length
        cons
          Text:     OK (0.38s)
            549  μs ±  43 μs, 20% faster than baseline
          LazyText: OK (0.36s)
            556  μs ±  27 μs, 17% faster than baseline
        decode
          Text:     OK (1.77s)
            776  μs ±  23 μs, 26% faster than baseline
          LazyText: OK (1.15s)
            1.01 ms ±  95 μs, 16% faster than baseline
        drop
          Text:     OK (0.21s)
            554  μs ±  54 μs, 16% faster than baseline
          LazyText: OK (0.19s)
            540  μs ±  46 μs, 21% faster than baseline
        filter
          Text:     OK (0.27s)
            7.39 ms ± 371 μs,  9% faster than baseline
          LazyText: OK (0.26s)
            7.72 ms ± 371 μs,  7% faster than baseline
        filter.filter
          Text:     OK (0.27s)
            7.32 ms ± 435 μs, 14% faster than baseline
          LazyText: OK (0.14s)
            7.60 ms ± 752 μs
        init
          Text:     OK (0.30s)
            518  μs ±  23 μs, 18% faster than baseline
          LazyText: OK (0.35s)
            542  μs ±  31 μs, 17% faster than baseline
        intercalate
          Text:     OK (0.34s)
            1.11 ms ±  62 μs, 21% faster than baseline
          LazyText: OK (0.37s)
            1.19 ms ±  84 μs, 20% faster than baseline
        intersperse
          Text:     OK (0.20s)
            532  μs ±  52 μs, 14% faster than baseline
          LazyText: OK (0.21s)
            545  μs ±  51 μs, 11% faster than baseline
        map
          Text:     OK (0.35s)
            527  μs ±  31 μs, 24% faster than baseline
          LazyText: OK (0.64s)
            536  μs ±  13 μs, 16% faster than baseline
        map.map
          Text:     OK (0.34s)
            499  μs ±  31 μs, 32% faster than baseline
          LazyText: OK (4.54s)
            538  μs ±  27 μs, 24% faster than baseline
        replicate char
          Text:     OK (0.36s)
            19.9 ns ± 1.7 ns, 23% faster than baseline
          LazyText: OK (0.46s)
            15.7 ns ± 1.2 ns, 22% faster than baseline
        replicate string
          Text:     OK (0.57s)
            22.6 ns ± 1.2 ns, 23% faster than baseline
          LazyText: OK (0.51s)
            18.4 ns ± 1.6 ns, 24% faster than baseline
        take
          Text:     OK (0.24s)
            332  μs ±  22 μs, 30% faster than baseline
          LazyText: OK (0.25s)
            346  μs ±  27 μs, 26% faster than baseline
        tail
          Text:     OK (0.20s)
            528  μs ±  47 μs, 24% faster than baseline
          LazyText: OK (0.34s)
            506  μs ±  38 μs, 28% faster than baseline
        toLower
          Text:     OK (0.21s)
            65.0 ms ± 3.4 ms, 18% faster than baseline
          LazyText: OK (0.28s)
            88.3 ms ± 5.3 ms, 17% faster than baseline
        toUpper
          Text:     OK (0.28s)
            89.2 ms ± 3.2 ms, 16% faster than baseline
          LazyText: OK (0.35s)
            111  ms ± 9.1 ms, 15% faster than baseline
        words
          Text:     OK (0.51s)
            15.1 ms ± 531 μs, 17% faster than baseline
          LazyText: OK (0.25s)
            31.7 ms ± 1.9 ms, 18% faster than baseline
        zipWith
          Text:     OK (0.33s)
            503  μs ±  30 μs, 29% faster than baseline
          LazyText: OK (0.34s)
            518  μs ±  52 μs, 27% faster than baseline
    russian
      length
        cons
          Text:     OK (0.24s)
            1.48 μs ±  93 ns, 29% faster than baseline
          LazyText: OK (0.24s)
            1.49 μs ± 102 ns, 29% faster than baseline
        decode
          Text:     OK (0.49s)
            3.20 μs ± 291 ns, 25% faster than baseline
          LazyText: OK (0.27s)
            3.53 μs ± 228 ns, 19% faster than baseline
        drop
          Text:     OK (0.26s)
            1.66 μs ± 112 ns, 14% faster than baseline
          LazyText: OK (0.25s)
            1.67 μs ± 100 ns, 14% faster than baseline
        filter
          Text:     OK (0.20s)
            22.3 μs ± 1.6 μs
          LazyText: OK (0.25s)
            26.8 μs ± 1.8 μs
        filter.filter
          Text:     OK (0.20s)
            22.1 μs ± 1.8 μs
          LazyText: OK (0.24s)
            26.8 μs ± 1.8 μs
        init
          Text:     OK (0.23s)
            1.57 μs ±  95 ns, 17% faster than baseline
          LazyText: OK (0.25s)
            1.59 μs ± 149 ns, 18% faster than baseline
        intercalate
          Text:     OK (0.20s)
            2.46 μs ± 232 ns, 25% faster than baseline
          LazyText: OK (0.20s)
            2.59 μs ± 254 ns, 41% faster than baseline
        intersperse
          Text:     OK (0.24s)
            1.61 μs ± 100 ns, 39% faster than baseline
          LazyText: OK (0.25s)
            1.58 μs ±  89 ns, 17% faster than baseline
        map
          Text:     OK (0.25s)
            1.57 μs ±  99 ns, 18% faster than baseline
          LazyText: OK (0.25s)
            1.58 μs ± 107 ns, 17% faster than baseline
        map.map
          Text:     OK (0.25s)
            1.59 μs ±  87 ns, 16% faster than baseline
          LazyText: OK (0.25s)
            1.58 μs ±  91 ns, 17% faster than baseline
        replicate char
          Text:     OK (0.22s)
            18.6 ns ± 1.4 ns, 18% faster than baseline
          LazyText: OK (0.32s)
            16.1 ns ± 780 ps, 14% faster than baseline
        replicate string
          Text:     OK (0.26s)
            23.2 ns ± 1.8 ns, 13% faster than baseline
          LazyText: OK (0.37s)
            18.0 ns ± 716 ps, 20% faster than baseline
        take
          Text:     OK (0.59s)
            1.04 μs ±  71 ns, 20% faster than baseline
          LazyText: OK (0.30s)
            1.04 μs ±  96 ns, 21% faster than baseline
        tail
          Text:     OK (0.25s)
            1.60 μs ± 150 ns, 16% faster than baseline
          LazyText: OK (0.25s)
            1.61 μs ± 147 ns, 14% faster than baseline
        toLower
          Text:     OK (0.26s)
            112  μs ±  11 μs
          LazyText: OK (0.17s)
            158  μs ±  13 μs
        toUpper
          Text:     OK (0.18s)
            150  μs ±  14 μs
          LazyText: OK (0.23s)
            201  μs ±  15 μs
        words
          Text:     OK (0.17s)
            33.2 μs ± 3.0 μs, 12% faster than baseline
          LazyText: OK (0.31s)
            67.7 μs ± 4.0 μs, 11% faster than baseline
        zipWith
          Text:     OK (0.22s)
            1.48 μs ±  82 ns, 24% faster than baseline
          LazyText: OK (0.24s)
            1.47 μs ±  92 ns, 25% faster than baseline
    japanese
      length
        cons
          Text:     OK (0.25s)
            1.61 μs ± 129 ns, 21% faster than baseline
          LazyText: OK (0.47s)
            1.61 μs ± 112 ns, 20% faster than baseline
        decode
          Text:     OK (0.97s)
            3.41 μs ±  54 ns, 19% faster than baseline
          LazyText: OK (0.54s)
            3.53 μs ± 243 ns, 22% faster than baseline
        drop
          Text:     OK (0.26s)
            1.65 μs ±  82 ns, 22% faster than baseline
          LazyText: OK (0.24s)
            1.67 μs ± 160 ns, 20% faster than baseline
        filter
          Text:     OK (0.20s)
            11.6 μs ± 1.0 μs, 11% faster than baseline
          LazyText: OK (0.14s)
            14.7 μs ± 1.4 μs, 14% faster than baseline
        filter.filter
          Text:     OK (0.43s)
            11.9 μs ± 340 ns, 12% faster than baseline
          LazyText: OK (0.53s)
            15.0 μs ± 1.0 μs, 14% faster than baseline
        init
          Text:     OK (0.24s)
            1.62 μs ±  84 ns, 22% faster than baseline
          LazyText: OK (0.51s)
            1.69 μs ± 132 ns, 18% faster than baseline
        intercalate
          Text:     OK (0.26s)
            3.36 μs ± 216 ns, 22% faster than baseline
          LazyText: OK (0.30s)
            3.93 μs ± 166 ns, 18% faster than baseline
        intersperse
          Text:     OK (0.25s)
            1.60 μs ± 103 ns, 22% faster than baseline
          LazyText: OK (0.26s)
            1.61 μs ± 113 ns, 21% faster than baseline
        map
          Text:     OK (0.25s)
            1.58 μs ± 145 ns, 22% faster than baseline
          LazyText: OK (0.25s)
            1.59 μs ± 145 ns, 21% faster than baseline
        map.map
          Text:     OK (0.25s)
            1.63 μs ± 123 ns, 17% faster than baseline
          LazyText: OK (0.26s)
            1.61 μs ± 121 ns, 16% faster than baseline
        replicate char
          Text:     OK (0.22s)
            19.4 ns ± 1.6 ns, 10% faster than baseline
          LazyText: OK (0.32s)
            15.1 ns ± 1.2 ns, 14% faster than baseline
        replicate string
          Text:     OK (0.25s)
            22.9 ns ± 1.7 ns, 13% faster than baseline
          LazyText: OK (0.22s)
            18.8 ns ± 1.3 ns, 15% faster than baseline
        take
          Text:     OK (0.19s)
            1.14 μs ± 102 ns, 21% faster than baseline
          LazyText: OK (0.18s)
            1.13 μs ±  93 ns, 21% faster than baseline
        tail
          Text:     OK (0.26s)
            1.72 μs ± 117 ns, 10% faster than baseline
          LazyText: OK (0.28s)
            1.81 μs ± 130 ns, 15% faster than baseline
        toLower
          Text:     OK (0.19s)
            74.9 μs ± 5.9 μs
          LazyText: OK (0.24s)
            105  μs ± 7.2 μs
        toUpper
          Text:     OK (0.17s)
            70.0 μs ± 5.4 μs,  7% faster than baseline
          LazyText: OK (0.22s)
            98.4 μs ± 6.4 μs, 15% faster than baseline
        words
          Text:     OK (0.21s)
            47.7 μs ± 4.4 μs, 21% faster than baseline
          LazyText: OK (0.21s)
            43.0 μs ± 3.9 μs, 10% faster than baseline
        zipWith
          Text:     OK (0.27s)
            1.69 μs ±  92 ns, 19% faster than baseline
          LazyText: OK (0.27s)
            1.73 μs ± 134 ns, 24% faster than baseline

All 216 tests passed (108.18s)
Benchmark text-benchmarks: FINISH
```

wz1000 commented:
I've successfully validated the implementation of `popcount16` using the following program:

```c
#include <stdint.h>
#include <stdio.h>
#include <assert.h>
#include <stdlib.h>
#include <sys/types.h>

static inline const size_t popcount16(uint16_t x) {

  // Taken from https://en.wikipedia.org/wiki/Hamming_weight
  const uint16_t m1  = 0x5555; //binary: 0101...
  const uint16_t m2  = 0x3333; //binary: 00110011..
  const uint16_t m4  = 0x0f0f; //binary:  4 zeros,  4 ones ...
  x -= (x >> 1) & m1;             //put count of each 2 bits into those 2 bits
  x = (x & m2) + ((x >> 2) & m2); //put count of each 4 bits into those 4 bits 
  x = (x + (x >> 4)) & m4;        //put count of each 8 bits into those 8 bits 
  return (x >> 8) + (x & 0x00FF);
}

int main() {
  for(int i = 0; i <= 0xFFFF; i++) {
    size_t a,b;
    a = __builtin_popcount((uint16_t) i);
    b = popcount16(i);
    if (a != b) {
      printf("No matches %d %d \n", a, b);
      exit(1);
    }
  }
  printf("All values validated\n");
}
```

wz1000 commented on tests/THLinkTest.hs:
```diff
@@ -0,0 +1,12 @@
+-- Simple test script for #450
```
I've done this.

Bodigrim commented:
Thanks @wz1000!

wz1000 commented:
@Bodigrim Could we have a release for inclusion into 9.4?

Bodigrim commented:
@wz1000 sure, I'm waiting for #448 to land into `master` before releasing.

Bodigrim commented:
@wz1000 @bgamari actually what is the timeline for GHC 9.4.1? Do we have any time left to finish #448? If no, could you please confirm that `master` branch works for GHC purposes as is?

wz1000 commented:
The RC will be out this week, but we can use the master branch for that without waiting for a release. The final release will be made in about 2 weeks (early August) and we do need a release by then, ideally before.

wz1000 commented:
I will test the master branch one more time to be sure, but I think all the patches we need have been merged.

bgamari commented:
@Bodigrim, I'm afraid 9.4 is essentially done. rc1 should be released by the end of today and will ship with 9412f442c36cf7d8c0690f7db4615f9288bd0ca7, which appears to be working well. It would be great if we could produce a `text` release from it or something closely related

Bodigrim commented:
@wz1000 @bgamari Released as [`text-2.0.1`](https://hackage.haskell.org/package/text-2.0.1), fdb06ff327519f3c0fc6cc9997b7cb7fe8ab8178.
